crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -288,41 +209,706 @@ impl CoinReadApiServer for CoinReadApi {
+            );
+        }
+
+        #[tokio::test]
+        async fn test_object_not_found() {
+            let owner = get_test_owner();
+            let object_id = ObjectID::random();
+            let mut mock_state = MockStateServiceWrapper::new();
+            mock_state.expect_get_object().returning(move |_| Ok(None));
+
+            let internal = CoinReadInternalService {
+                state: Arc::new(mock_state),
+                metrics: Arc::new(JsonRpcMetrics::new_for_tests()),
+            };
+
+            let coin_read_api = CoinReadApi {
+                internal: Arc::new(internal),
+            };
+
+            let response = coin_read_api
+                .get_all_coins(owner, Some(object_id), None)
+                .await;
+
+            assert!(response.is_err());
+            let error_result = response.unwrap_err();
+            let error_object: ErrorObjectOwned = error_result.into();
+            assert_eq!(error_object.code(), -32602);
+            assert_eq!(
+                error_object.message(),
+                format!("Invalid Cursor {:?}, Object not found", object_id)
+            );
+        }
+    }
+
+    mod get_balance_tests {
+        use super::super::*;
+        use super::*;
+        use jsonrpsee::types::ErrorObjectOwned;
+
+        #[tokio::test]
+        async fn test_get_balance_index_store_not_available() {
+            let owner = get_test_owner();
+            let coin_type = get_test_coin_type(get_test_package_id());
+            let mut mock_state = MockStateServiceWrapper::new();
+            mock_state
+                .expect_get_balance()
+                .returning(move |_, _| Err(SuiError::IndexStoreNotAvailable));
+            let internal = CoinReadInternalService {
+                state: Arc::new(mock_state),
+                metrics: Arc::new(JsonRpcMetrics::new_for_tests()),
+            };
+            let coin_read_api = CoinReadApi {
+                internal: Arc::new(internal),
+            };
+
+            let response = coin_read_api
+                .get_balance(owner, Some(coin_type.to_string()))
+                .await;
+
+            assert!(response.is_err());
+            let error_result = response.unwrap_err();
+            let error_object: ErrorObjectOwned = error_result.into();
+            assert_eq!(error_object.code(), -32000);
+            assert_eq!(
+                error_object.message(),
+                "Index store not available on this Fullnode."
+            );
+        }
+
+        #[tokio::test]
+        async fn test_get_balance_execution_error() {
+            let owner = get_test_owner();
+            let coin_type = get_test_coin_type(get_test_package_id());
+            let mut mock_state = MockStateServiceWrapper::new();
+            mock_state
+                .expect_get_balance()
+                .returning(move |_, _| Err(SuiError::ExecutionError("mock db error".to_string())));
```

Comment:
For these mock errors, can we document what real situation(s) can cause this kind of error?  I can't intuitively think what `get_balance` would be executing.

Reply:
Yes for sure! And I think this specific example harkens back to our discussion about whether it makes sense to mock internal server errors. In this scenario, indeed it's not clear what would need to occur to manifest `SuiError::ExecutionError`, and mocking such an error value might be too far removed to be useful. My intention was to validate that errors upstream are surfaced in a legible manner, but perhaps these errors are better served living in `sui-json-rpc/src/error.rs` instead of within the specific api, since the intention is no longer, "Do we return the expected error for a given error scenario" and instead, "Is the returned error meaningful"

Reply:
Got it -- in that case, I would say that we should have two tests:

- One to test forwarding an expected error (for which we should only use errors that we expect).
- One to test what happens when we get an unexpected error (which should translate into an internal server error that we set-up logging and alerting for).

Reply:
Just re-surfacing this comment (I think I may have commented on it after it was resolved, but with an action item in my previous message, which wasn't helpful of me).

Reply:
Each rpc method now has success scenarios covering all possibilities, all expected scenarios, and one example of an unexpected error scenario